pkg/strategy/dca2/recover.go

```diff
@@ -0,0 +1,268 @@
+
+	// recover state
+	state, err := recoverState(ctx, s.Symbol, s.Short, int(s.MaxOrderNum), openOrders, currentRound, s.OrderExecutor.ActiveMakerOrders(), s.OrderExecutor.OrderStore(), s.OrderGroupID)
+	if err != nil {
+		return err
+	}
+
+	// recover position
+	if err := recoverPosition(ctx, s.Position, queryService, currentRound); err != nil {
+		return err
+	}
+
+	// recover budget
+	budget := recoverBudget(currentRound)
+
+	// recover startTimeOfNextRound
+	startTimeOfNextRound := recoverStartTimeOfNextRound(ctx, currentRound, s.CoolDownInterval)
+
+	s.state = state
+	if !budget.IsZero() {
+		s.Budget = budget
+	}
+	s.startTimeOfNextRound = startTimeOfNextRound
+
+	return nil
+}
+
+// recover state
+func recoverState(ctx context.Context, symbol string, short bool, maxOrderNum int, openOrders []types.Order, currentRound Round, activeOrderBook *bbgo.ActiveOrderBook, orderStore *core.OrderStore, groupID uint32) (State, error) {
+	numOpenOrders := len(openOrders)
+	// dca stop at take profit order stage
+	if currentRound.TakeProfitOrder.OrderID != 0 {
+		// check the open orders is take profit order or not
+		if numOpenOrders == 1 {
+			if openOrders[0].OrderID == currentRound.TakeProfitOrder.OrderID {
+				activeOrderBook.Add(openOrders[0])
+				// current round's take-profit order still opened, wait to fill
+				return TakeProfitReady, nil
+			} else {
+				return None, fmt.Errorf("stop at taking profit stage, but the open order's OrderID is not the take-profit order's OrderID")
+			}
+		}
+
+		if numOpenOrders == 0 {
+			// current round's take-profit order filled, wait to open next round
+			return WaitToOpenPosition, nil
+		}
+
+		return None, fmt.Errorf("stop at taking profit stage, but the number of open orders is > 1")
+	}
+
+	if len(currentRound.OpenPositionOrders) == 0 {
+		// new strategy
+		return WaitToOpenPosition, nil
+	}
+
+	numOpenPositionOrders := len(currentRound.OpenPositionOrders)
+	if numOpenPositionOrders > maxOrderNum {
+		return None, fmt.Errorf("the number of open-position orders is > max order number")
+	} else if numOpenPositionOrders < maxOrderNum {
+		// failed to place some orders at open position stage
+		return None, fmt.Errorf("the number of open-position orders is < max order number")
+	}
+
+	if numOpenOrders > numOpenPositionOrders {
+		return None, fmt.Errorf("the number of open orders is > the number of open-position orders")
+	}
+
+	if numOpenOrders == numOpenPositionOrders {
+		activeOrderBook.Add(openOrders...)
+		orderStore.Add(openOrders...)
+		return OpenPositionReady, nil
+	}
+
+	var openedCnt, filledCnt, cancelledCnt int64
+	for _, order := range currentRound.OpenPositionOrders {
+		switch order.Status {
+		case types.OrderStatusNew, types.OrderStatusPartiallyFilled:
+			openedCnt++
+		case types.OrderStatusFilled:
+			filledCnt++
+		case types.OrderStatusCanceled:
+			cancelledCnt++
+		default:
+			return None, fmt.Errorf("there is unexpected status %s of order %s", order.Status, order)
+		}
+	}
+
```

Comment:
Could I say `filledCnt == 0` is an error case at this point?

Reply:
Not really....
Because we place the open-position orders first with the best ask but the best ask order may be filled before we place it.

Reply:
He filtered the orders with executedQuantity > 0